### PR TITLE
[FLINK-19626][table-planner-blink] Introduce multi-input operator construction algorithm

### PR DESCRIPTION
**Pull request by TsReaper**
## What is the purpose of the change

As multiple input exec nodes have been introduced, we're going to construct multiple input operators by a new construction algorithm. This PR introduces such algorithm.

Multiple input optimization is currently not use by default and is only used for tests as its operator is not ready. We'll change this once the multiple input operator is ready.

## Brief change log

 - Introduce multi-input operator construction algorithm

## Verifying this change

This change added tests and can be verified as follows: Run the newly added tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/MultipleInputNodeCreationProcessor.java [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+		for (ExecNodeWrapper inputWrapper : wrapper.inputs) {
+			if (inputWrapper.group == wrapper.group) {
+				// one of the input is in the same group, so this node is not the tail of the group
+				return false;
+			}
+		}
+		return true;
+	}
+
+	private boolean isNewSource(ExecNode<?, ?> node) {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/MultipleInputNodeCreationProcessor.java:306]: `SourceProvider` could also provide new `Source`

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/MultipleInputNodeCreationProcessor.java [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+					((BatchExecExchange) inputWrapper.execNode)
+						.distribution.getType() == RelDistribution.Type.SINGLETON);
+
+			if (shouldRemove) {
+				wrapper.removeFromGroup();
+			}
+		}
+	}
+
+	private boolean isTailOfMultipleInputGroup(ExecNodeWrapper wrapper) {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/MultipleInputNodeCreationProcessor.java:293]: `Tail` => `Header` ?

**Review by godfreyhe — COMMENTED** [review]
Thanks for the contribute @TsReaper

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/MultipleInputNodeCreationProcessor.java [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+			this.inputs = new ArrayList<>();
+			this.outputs = new ArrayList<>();
+			this.group = null;
+		}
+
+		private void createGroup() {
+			this.group = new MultipleInputGroup(this);
+		}
+
+		private void addToGroup(MultipleInputGroup group) {
```
[the remark this answers is not in the document]
Comment (TsReaper) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/MultipleInputNodeCreationProcessor.java:454]: We expect this wrapper not to be in any group.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/MultipleInputNodeCreationProcessor.java [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+				ShuffleMode.PIPELINED);
+			resolver.detectAndResolve();
+		}
+
+		List<ExecNodeWrapper> sinkWrappers = wrapExecNodes(sinkNodes);
+		// sort all nodes in topological order, sinks come first and sources come last
+		List<ExecNodeWrapper> orderedWrappers = topologicalSort(sinkWrappers);
+		// group nodes into multiple input groups
+		createMultipleInputGroups(orderedWrappers);
+		// apply optimizations to remove unnecessary nodes out of multiple input groups
```
[the remark this answers is not in the document]
Comment (TsReaper) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/MultipleInputNodeCreationProcessor.java:88]: They're best-to-have, not a must.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/MultipleInputNodeCreationProcessor.java (hunk 1 above)
Comment (TsReaper) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/MultipleInputNodeCreationProcessor.java:306]: FLIP-146 is not ready when this PR is submitted. I'll rebase the master branch.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/MultipleInputNodeCreationProcessor.java (hunk 2 above)
Comment (TsReaper) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/MultipleInputNodeCreationProcessor.java:293]: I actually don't like the idea of `head` and `tail` in the multiple input operator as they're ambiguous. Why don't we use phrases like `input`, `entrance` and `root`?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/batch/BatchExecMultipleInputNode.scala [hunk 5]
```diff
@@ -99,6 +100,14 @@ class BatchExecMultipleInputNode(
     val memoryKB = generator.getManagedMemoryWeight
     ExecNode.setManagedMemoryWeight(multipleInputTransform, memoryKB * 1024)
 
+    if (withSourceChaining) {
+      // set chaining strategy for source chaining
+      multipleInputTransform.setChainingStrategy(ChainingStrategy.HEAD_WITH_SOURCES)
```
[the remark this answers is not in the document]
Comment (TsReaper) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/batch/BatchExecMultipleInputNode.scala:105]: It's OK. I just want to be more precise.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/processor/utils/TopologyGraph.java [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+		fromNode.outputs.remove(toNode);
+		toNode.inputs.remove(fromNode);
+	}
+
+	/**
+	 * Calculate the maximum distance of the currently added nodes from the nodes without inputs.
+	 * The smallest distance is 0 (which are exactly the nodes without inputs) and the distances of
+	 * other nodes are the largest distances in their inputs plus 1.
+	 */
+	Map<ExecNode<?, ?>, Integer> calculateDistance() {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/processor/utils/TopologyGraph.java:104]: give `distance` a definition ?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/processor/utils/TopologyGraph.java [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+				}
+				visited.add(next);
+				queue.offer(next);
+			}
+		}
+
+		return false;
+	}
+
+	private TopologyNode getTopologyNode(ExecNode<?, ?> execNode) {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/processor/utils/TopologyGraph.java:170]: getOrCreateTopologyNode ?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/processors/utils/TopologyGraph.java [hunk 8]
```diff
@@ -100,6 +100,9 @@ void unlink(ExecNode<?, ?> from, ExecNode<?, ?> to) {
 	 * Calculate the maximum distance of the currently added nodes from the nodes without inputs.
 	 * The smallest distance is 0 (which are exactly the nodes without inputs) and the distances of
 	 * other nodes are the largest distances in their inputs plus 1.
+	 *
+	 * <p>Distance of a node is defined as the number of edges one needs to go through from the
+	 * nodes without inputs to this node.
 	 */
 	Map<ExecNode<?, ?>, Integer> calculateDistance() {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/processors/utils/TopologyGraph.java]: rename to calculateMaxDistance

**Comment by TsReaper** [comment]
Azure passed in https://dev.azure.com/tsreaper96/Flink/_build/results?buildId=100&view=results

**Review by godfreyhe — APPROVED** [review]
LGTM, +1 to merge

**Pull request merged**